packages/react-is/src/ReactIs.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+'use strict';
+
+import {
+  REACT_ASYNC_MODE_TYPE,
+  REACT_CONTEXT_TYPE,
+  REACT_ELEMENT_TYPE,
+  REACT_FRAGMENT_TYPE,
+  REACT_PORTAL_TYPE,
+  REACT_PROVIDER_TYPE,
+  REACT_STRICT_MODE_TYPE,
+} from 'shared/ReactSymbols';
+
+// e.g. Fragments, StrictMode
+function getType(object: any) {
+  return typeof object === 'object' && object !== null ? object.type : null;
+}
+
+// e.g. Elements, Portals
+function getTypeOf(object: any) {
+  return typeof object === 'object' && object !== null ? object.$$typeof : null;
+}
+
+// e.g. Context provider and consumer
+function getTypeTypeOf(object: any) {
+  return typeof object === 'object' &&
+    object !== null &&
+    typeof object.type === 'object' &&
+    object.type !== null
+    ? object.type.$$typeof
+    : null;
+}
+
+export function typeOf(object: any) {
+  let maybeType = getType(object);
+  switch (maybeType) {
+    case REACT_ASYNC_MODE_TYPE:
+    case REACT_FRAGMENT_TYPE:
+    case REACT_STRICT_MODE_TYPE:
+      if (getTypeOf(object) === REACT_ELEMENT_TYPE) {
+        return maybeType;
+      }
+  }
+
+  maybeType = getTypeTypeOf(object);
+  switch (maybeType) {
+    case REACT_CONTEXT_TYPE:
+    case REACT_PROVIDER_TYPE:
+      return maybeType;
+  }
+
+  maybeType = getTypeOf(object);
+  switch (maybeType) {
+    case REACT_ELEMENT_TYPE:
+    case REACT_PORTAL_TYPE:
+      return maybeType;
+  }
+}
+
+export const AsyncMode = REACT_ASYNC_MODE_TYPE;
+export const ContextConsumer = REACT_CONTEXT_TYPE;
+export const ContextProvider = REACT_PROVIDER_TYPE;
+export const Element = REACT_ELEMENT_TYPE;
+export const Fragment = REACT_FRAGMENT_TYPE;
+export const Portal = REACT_PORTAL_TYPE;
+export const StrictMode = REACT_STRICT_MODE_TYPE;
+
+export function isAsyncMode(object: any) {
+  return (
```

Comment:
Seems like all of these could be rewritten as `getTypeOf` comparisons to ensure 1:1 check parity?

Reply:
There is a lot of back and forth on a PR, and the latest round seems entirely like nits. Seems a bit inefficient.

WRT using typeOf here- Either way should be fine. It's just a subjective preference, right?

Reply:
PS using typeOf in isElement didn't work, because the return type was overly specific. I didn't feel strongly about it, but I opted out of using typeOf director for that reason.

Reply:
I don't agree that repeating the same checks many times is a nit. Generally we avoid this everywhere else in the codebase. (If you're referring to the previous comment.)

Yes, they probably don't make a difference in a large scheme of things, but if we can make it more efficient, why not do it?

For this comment, I'm not sure. The issue I noted in https://github.com/facebook/react/pull/12199#discussion_r167444125i was a correctness issue. It was fixed in https://github.com/facebook/react/pull/12199/commits/b22a7b78836d7d8252989c7a9a7522c40e7d2637. The fix required changes in two places. My concern is that it's confusing to remember both those places need to contain this logic. When we add a new type it won't be obvious which ones of those `get*` functions need to be called for `is*` to work consistently. Expressing this via `typeof()` solves this problem by consolidating that logic in one place.

Reply:
>PS using typeOf in isElement didn't work, because the return type was overly specific. I didn't feel strongly about it, but I opted out of using typeOf director for that reason.

You probably considered it but another way to express this would be

```js
switch (typeof(obj)) {
  case Element:
  case ContextConsumer:
  case ContextProvider:
    return true;
  default:
    return false;
}
```

I agree there's no clear winner in this particular case though.

Reply:
> Yes, they probably don't make a difference in a large scheme of things, but if we can make it more efficient, why not do it?

I don't think the specific comment this is in response to is really more efficient one way or another, in terms of runtime efficiency. And the complexity/code-size difference is minimal. This is why I said it felt like a nit.

Reply:
See my response below. Again I agree it probably doesn't matter in practice. But since it's a core utility and it's easy for us to minimize property access I don't see a reason not to.

Reply:
Sure, the comment we're responding to is the implementation of `isAsyncMode`, so the runtime path of `typeOf` didn't seem relevant since it wasn't being used (just the helper getter functions).

It's fine. I'm about to push the change you suggested.